src/query_plan/mod.rs

```diff
@@ -1 +1,210 @@
+use apollo_compiler::executable::{Field, InlineFragment, OperationType, Selection, SelectionSet};
+use apollo_compiler::{ExecutableDocument, NodeStr};
+use std::sync::Arc;
+
 pub mod operation;
+
+pub struct QueryPlan {
+    node: Option<TopLevelPlanNode>,
+}
+
+pub enum TopLevelPlanNode {
+    Subscription(SubscriptionNode),
+    Fetch(FetchNode),
+    Sequence(SequenceNode),
+    Parallel(ParallelNode),
+    Flatten(FlattenNode),
+    Defer(DeferNode),
+    Condition(ConditionNode),
+}
+
+pub struct SubscriptionNode {
+    primary: FetchNode,
+    rest: Option<PlanNode>,
+}
+
+pub enum PlanNode {
+    Fetch(Arc<FetchNode>),
+    Sequence(Arc<SequenceNode>),
+    Parallel(Arc<ParallelNode>),
+    Flatten(Arc<FlattenNode>),
+    Defer(Arc<DeferNode>),
+    Condition(Arc<ConditionNode>),
+}
+
+pub struct FetchNode {
+    service_name: NodeStr,
+    /// Optional identifier for the fetch for defer support. All fetches of a given plan will be
+    /// guaranteed to have a unique `id`.
+    id: Option<NodeStr>,
+    /// If query planner defer support is enabled _and_ the subgrpah named `service_name` supports
+    /// defer, then this boolean says whether `operation` contains some @defer. Unset otherwise.
+    has_defers: Option<bool>,
+    variable_usages: Vec<NodeStr>,
```

Comment:
should this be a type, just like `requires` below is a `Selection`? I take it this is the variables in arguments, i.e. `Value` type. Although I do understand having it be a `Value` adds a certain level of complexity of working with an enum, when we are only ever interested in the one variant - `Value::Variable`.

Reply:
I think if it's only variable names the `Value` enum is a bit much, but it could use `Name` instead of `NodeStr`

Reply:
It's indeed just variable names, changing it to `Name`s should be better.